src/debug_training/ROS_Tools.md

```diff
@@ -0,0 +1,316 @@
+# WRoverPlayground - ROS Tools
+
+## ROS Concepts Crash Course
+
+ROS nodes are processes that use ROS to communicate with other nodes.
+ROS nodes communicate with each other using streaming topics, RPC services, and the Parameter Server.
+
+ROS topics are named interfaces with a publisher and subscriber model.
+Publisher nodes generate data and publish that data to a topic.
+Subscriber nodes subscribe to a topic to use data that is published to a topic.
+A topic can have multiple publisher and subscriber nodes.
+Topics are intended for unidirectional communication.
+If a response is expected from a request, a service should be used instead.
+
+ROS services allow nodes to return data to other nodes.
+Server nodes define the service request and response.
+Client nodes call the service by sending a request and await the response.
+
+Both topics and services are defined using messages.
+A message is a data structure containing typed fields.
+For a service, a request and a response message must be defined.
+
+ROS topics and services will be covered in more detail in the next training modules.
+
+The final method of communication between ROS nodes is the Parameter Service.
+This is essentially a shared dictionary that can be used to store and retrieve variables at runtime.
+It is usually used for configuration since it is not designed for high performance.
+
+## ROS Command Line
+
+### Starting nodes
+
+Open a terminal window and run `roscore`.
+This starts the ROS master node.
+
+Open a new terminal and run `rosnode list`.
+You should see a single node: `/rosout`.
+
+Run `rosnode info /rosout`.
+You should see the following output: 
+```
+Node [/rosout]
+Publications: 
+ * /rosout_agg [rosgraph_msgs/Log]
+
+Subscriptions: 
+ * /rosout [unknown type]
+
+Services: 
+ * /rosout/get_loggers
+ * /rosout/set_logger_level
+```
+This shows that the `/rosout` node publishes to the `/rosout_agg` topic and subscribes sto the `/rosout` topic.
+Also, the `/rosout` node provides the `/rosout/get_loggers` and `/rosout/set_logger_level` services.
+
+Now, let's start another node.
+Run `rosrun turtlesim turtlesim_node`.
+You should see a GUI window with a turtle in the middle.
+
+![TurtleSim GUI](./images/turtlesim_1.png)
+
+In a new terminal, run `rosnode list`.
+You should see the following output: 
+```
+/rosout
+/turtlesim
+```
+
+Let's find out more about the `/turtlesim` node.
+Run `rosnode info /turtlesim`.
+You should see the following output:
+```
+Node [/turtlesim]
+Publications: 
+ * /rosout [rosgraph_msgs/Log]
+ * /turtle1/color_sensor [turtlesim/Color]
+ * /turtle1/pose [turtlesim/Pose]
+
+Subscriptions: 
+ * /turtle1/cmd_vel [unknown type]
+
+Services: 
+ * /clear
+ * /kill
+ * /reset
+ * /spawn
+ * /turtle1/set_pen
+ * /turtle1/teleport_absolute
+ * /turtle1/teleport_relative
+ * /turtlesim/get_loggers
+ * /turtlesim/set_logger_level
+```
```

Comment:
Maybe add a short description of what they are seeing. Like, "As you see, for `/turtlesim` . . ." Don't explain what everything is. Just include a sentence or two to connect the output from the previous run of `rosnode info` to this run.